Twitter format strings

WPettersson opened the pull request:
- Allow users to specify how tweets should be displayed

3 new settings are available. twitter_format_string sets how tweets should be
displayed, with variables allowing for insertion of appropriate content. These
variables are outlined further below. If a retweet or reply is detected, the
appropriate extra format string(s) is(are) expanded and then inserted into
either %r or %t. The default values keep the same display as before, and are:

twitter_format_string = "\002[\002%i%r\002]\002 %t%c"
twitter_retweet_format_string = "RT %a: "
twitter_reply_format_string = "->%t"

twitter_format_string has the following 4 variables
%i - id (2 character hex) of the tweet
%r - reply format (see below)
%t - retweet format (see below)
%c - content of the actual tweet

twitter_retweet_format_string only has one variable
%r - original author

twitter_reply_format_string also only has one variable
%i - id (2 character hex) of tweet that is being replied to

WPettersson committed:
Twitter format strings

- Allow users to specify how tweets should be displayed

3 new settings are available. twitter_format_string sets how tweets should be
displayed, with variables allowing for insertion of appropriate content. These
variables are outlined further below. If a retweet or reply is detected, the
appropriate extra format string(s) is(are) expanded and then inserted into
either %r or %t. The default values keep the same display as before, and are:

twitter_format_string = "\002[\002%i%r\002]\002 %t%c"
twitter_retweet_format_string = "RT %a: "
twitter_reply_format_string = "->%t"

twitter_format_string has the following 4 variables
%i - id (2 character hex) of the tweet
%r - reply format (see below)
%t - retweet format (see below)
%c - content of the actual tweet

twitter_retweet_format_string only has one variable
%r - original author

twitter_reply_format_string also only has one variable
%i - id (2 character hex) of tweet that is being replied to

jgeboski commented on protocols/twitter/twitter_lib.c:
```diff
@@ -602,6 +604,90 @@ static gboolean twitter_xt_get_status_list(struct im_connection *ic, const json_
+			if (*c == '%' && *(c+1)) {
+				if (*(c+1) == 'i') {
+					gchar *id = g_strdup_printf("%02x", reply_to);
+					reply = g_string_append(reply, id);
```
Just use g_string_append_printf().

jgeboski commented on protocols/twitter/twitter_lib.c:
```diff
@@ -602,6 +604,90 @@ static gboolean twitter_xt_get_status_list(struct im_connection *ic, const json_
+				case 'i': ; // Extra semi-colon avoids compiler error
+				            // "declaration after label"
+					gchar * id = g_strdup_printf("%02x", log_id);
+					text = g_string_append(text, id);
```
Again, just use g_string_append_printf().

jgeboski commented:
Since the introduced settings are unique to a twitter account, it might be worth dropping the twitter_ prefix. The settings names are excessively long.

jgeboski commented on protocols/twitter/twitter_lib.c:
```diff
@@ -602,6 +604,90 @@ static gboolean twitter_xt_get_status_list(struct im_connection *ic, const json_
+	GString * text = g_string_new(NULL);
+	if (txs->rt) {
+		gchar * c = rt_format;
+		while (*c) {
```
Since `c` is *always* being incremented for every iteration of the loop, just use a for loop.  It is far easier to follow.

jgeboski commented on protocols/twitter/twitter_lib.c:
```diff
@@ -602,6 +604,90 @@ static gboolean twitter_xt_get_status_list(struct im_connection *ic, const json_
+
+	if (reply_to != -1) {
+		gchar * c = reply_format;
+		while (*c) {
```
Again, use a `for` loop.

jgeboski commented on protocols/twitter/twitter_lib.c:
```diff
@@ -602,6 +604,90 @@ static gboolean twitter_xt_get_status_list(struct im_connection *ic, const json_
+
+	gchar * c = format;
+	while (*c) {
+		if (*c == '%' && *(c+1)) {
```
Again, use a `for` loop.

jgeboski commented on protocols/twitter/twitter_lib.c:
```diff
@@ -602,6 +604,90 @@ static gboolean twitter_xt_get_status_list(struct im_connection *ic, const json_
+	}
+
+	gchar * c = format;
+	while (*c) {
```
Again, use a for loop.

jgeboski commented:
Several of the conditional statements within the loops should probably inverted with the usage of `continue` in order to reduce excessive nesting.

WPettersson committed:
Rename settings to avoid "twitter_" prefix

dequis commented on protocols/twitter/twitter_lib.c:
```diff
@@ -602,6 +604,90 @@ static gboolean twitter_xt_get_status_list(struct im_connection *ic, const json_
+
+	GString * rt = g_string_new(NULL);
+	GString * reply = g_string_new(NULL);
+	GString * text = g_string_new(NULL);
```
Not sure about these three. It's instantiating three GStrings for three different kinds of messages that don't apply to all tweets.

jgeboski commented:
Solid patch, solid idea.  Just need to address some knit-picks.

WPettersson committed:
Minor tweaks

- Add comment describing function
- GStrings rt and reply are only initialised and free'd if needed
- Move to for() loops rather than while()
- Only instantiate gchar *c once.
- Invert some conditionals and use continue to avoid excessive nesting
- Use g_string_append_printf directly

WPettersson committed:
Documentation for format_strings

WPettersson committed:
Rewrite function - simplify

- Each format_string is now completely distinct, no embedding
- Format string selected, then looped through
- Variables used in wrong string are skipped
- Fixed up prefix handling

dequis commented on protocols/twitter/twitter.c:
```diff
@@ -329,6 +329,10 @@ static void twitter_init(account_t * acc)
+
+	s = set_add(&acc->set, "format_string", "\002[\002%i%r\002]\002 %t%c", NULL, acc);
+	s = set_add(&acc->set, "retweet_format_string", "RT %a: ", NULL, acc);
+	s = set_add(&acc->set, "reply_format_string", "->%i", NULL, acc);
```
the default values are outdated

dequis commented on protocols/twitter/twitter_lib.c:
```diff
@@ -602,6 +604,49 @@ static gboolean twitter_xt_get_status_list(struct im_connection *ic, const json_
+			break;
+		case 'r':
+			if (reply_to != -1) // In case someone does put %r in the wrong format_string
+			g_string_append_printf(text, "%02x", reply_to);
```
indentation

dequis commented on protocols/twitter/twitter_lib.c:
```diff
@@ -482,9 +484,9 @@ static struct twitter_xml_status *twitter_xt_get_status(const json_value *node)
+			txs->text = g_strdup(rtxs->text);
 			txs->id = rtxs->id;
-			txs_free(rtxs);
+			txs->rt = rtxs;
```
this doesn't seem to get free()'d anywhere

dequis replied:
nevermind i'm blind

WPettersson committed:
Update default values

WPettersson committed:
Fix indentation

dequis commented:
Hey! We're in the process of moving bitlbee to an official git repo:

https://github.com/bitlbee/bitlbee/

If you could squash, rebase and resubmit this PR to that repo, that would be cool. You also need to fork that other repo, since apparently there's no way to send pull requests across different fork networks. I added some notes on how to rebase stuff to [the README of this repo][1]

If you don't have time or can't do that for whatever reason, I'll take care of it. This particular PR just needs some testing so it might land directly in the bitlbee/develop branch (to get merged in bitlbee/master later). I might also go ahead and do PRs for the other patches you have

[1]: https://github.com/dequis/bitlbee/blob/develop/README.md

WPettersson commented:
I'll fix up this PR and get it onto the other PR. I'll do the same for the other two as well, although the threaded one does conflict with this PR so I may put that on hold until this gets merged.

dequis commented:
Thanks!

---> https://github.com/bitlbee/bitlbee/pull/10
